data/transactions/logic/eval.go

```diff
@@ -3548,14 +3548,22 @@ func (cx *EvalContext) accountReference(account stackValue) (basics.Address, uin
 	idx, err := cx.txn.Txn.IndexByAddress(addr, cx.txn.Txn.Sender)
 
 	invalidIndex := uint64(len(cx.txn.Txn.Accounts) + 1)
-	// Allow an address for an app that was created in group
+	// Allow an address for an app that was created in group or provided
+	// in the foreign apps array.
 	if err != nil && cx.version >= createdResourcesVersion {
 		for _, appID := range cx.created.apps {
 			createdAddress := cx.getApplicationAddress(appID)
 			if addr == createdAddress {
 				return addr, invalidIndex, nil
 			}
 		}
+
+		for _, appID := range cx.txn.Txn.ForeignApps {
+			foreignAddress := cx.getApplicationAddress(appID)
+			if addr == foreignAddress {
+				return addr, invalidIndex, nil
+			}
+		}
```

Comment:
This block will need to be feature-glagged by a new version, rather than `createdResourcesVersion`. Probably something like `appAddressAvailableVersion`.